File: py/src/imbrace/resources/conversations.py
from typing import Any, Dict, List, Optional, Union
from ..http import HttpTransport, AsyncHttpTransport
from ..types.conversation import (
    Conversation, ConversationActionResponse, JoinConversationInput,
    LeaveConversationInput, UpdateStatusInput, UpdateNameInput,
    InitVideoCallInput, AssignTeamMemberInput, RemoveTeamMemberInput,
    InvitableUser, CreateConversationInput, JoinRequestInput
)
from ..types.common import PagedResponse

class ConversationsResource:
    """Conversations domain — Sync.

    @param base - channel-service base URL (gateway/channel-service)
    """

    def __init__(self, http: HttpTransport, base: str):
        self._http = http
        self._base = base.rstrip("/")

    @property
    def _v1(self) -> str:
        return f"{self._base}/v1"

    @property
    def _v2(self) -> str:
        return f"{self._base}/v2"

    def list(self, type: Optional[str] = None, q: Optional[str] = None,
             limit: Optional[int] = None, skip: Optional[int] = None,
             sort: Optional[str] = None) -> PagedResponse[Conversation]:
        """List team conversations (V2)."""
        params: Dict[str, Any] = {}
        if type: params["type"] = type
        if q: params["q"] = q
        if limit: params["limit"] = limit
        if skip is not None: params["skip"] = skip
        if sort: params["sort"] = sort

        return self._http.request("GET", f"{self._v2}/team_conversations", params=params).json()

    def get(self, conv_id: str) -> Conversation:
        return self._http.request("GET", f"{self._v1}/team_conversations/{conv_id}").json()

    def get_by_conversation_id(self, conversation_id: str) -> Conversation:
        params = {"type": "conversation_id", "q": conversation_id}
        return self._http.request("GET", f"{self._v1}/team_conversations", params=params).json()

    def search(self, business_unit_id: str, q: str,
               limit: Optional[int] = None, skip: Optional[int] = None) -> PagedResponse[Conversation]:
        params = {
            "business_unit_id": business_unit_id,
            "type": "text",
            "q": q,
        }
        if limit: params["limit"] = limit
        if skip is not None: params["skip"] = skip
        return self._http.request("GET", f"{self._v1}/team_conversations/_search", params=params).json()

    def get_views_count(self, type: Optional[str] = None, q: Optional[str] = None) -> Dict[str, int]:
        params = {}
        if type: params["type"] = type
        if q: params["q"] = q
        return self._http.request("GET", f"{self._v2}/team_conversations/_views_count", params=params).json()

    def get_outstanding(self, business_unit_id: str,
                        limit: Optional[int] = None, skip: Optional[int] = None) -> PagedResponse[Conversation]:
        params = {"type": "business_unit_id", "q": business_unit_id}
        if limit: params["limit"] = limit
        if skip is not None: params["skip"] = skip
        return self._http.request("GET", f"{self._v1}/team_conversations/_outstanding", params=params).json()

    def join(self, body: JoinConversationInput) -> ConversationActionResponse:
        return self._http.request("POST", f"{self._v1}/team_conversations/_join", json=body).json()

    def leave(self, body: LeaveConversationInput) -> ConversationActionResponse:
        return self._http.request("POST", f"{self._v1}/team_conversations/_leave", json=body).json()

    def update_status(self, body: UpdateStatusInput) -> ConversationActionResponse:
        return self._http.request("POST", f"{self._v1}/team_conversations/_update_status", json=body).json()

    def update_name(self, body: UpdateNameInput) -> ConversationActionResponse:
        return self._http.request("POST", f"{self._v1}/team_conversations/_update_name", json=body).json()

    def init_video_call(self, body: InitVideoCallInput) -> ConversationActionResponse:
        return self._http.request("POST", f"{self._v1}/team_conversations/_init_jaas_conference", json=body).json()

    def assign_team_member(self, body: AssignTeamMemberInput) -> ConversationActionResponse:
        return self._http.request("POST", f"{self._v1}/team_conversations/assign_team_member", json=body).json()

    def remove_team_member(self, body: RemoveTeamMemberInput) -> ConversationActionResponse:
        return self._http.request("POST", f"{self._v1}/team_conversations/remove_team_member", json=body).json()

    def get_invitable_users(self, team_conv_id: str) -> List[InvitableUser]:
        return self._http.request("GET", f"{self._v1}/team_conversations/{team_conv_id}/users").json()

    def get_conversation(self, conversation_id: str) -> Conversation:
        """Get base conversation details."""
        return self._http.request("GET", f"{self._v1}/conversations/{conversation_id}").json()

    def create(self, body: Optional[CreateConversationInput] = None) -> Conversation:
        return self._http.request("POST", f"{self._v1}/conversations", json=body or {}).json()

    def join_request(self, body: JoinRequestInput) -> ConversationActionResponse:
        return self._http.request("POST", f"{self._v1}/team_conversations/_join_request", json=body).json()


class AsyncConversationsResource:
    """Conversations domain — Async."""

    def __init__(self, http: AsyncHttpTransport, base: str):
        self._http = http
        self._base = base.rstrip("/")

    @property
    def _v1(self) -> str:
        return f"{self._base}/v1"

    @property
    def _v2(self) -> str:
        return f"{self._base}/v2"

    async def list(self, type: Optional[str] = None, q: Optional[str] = None,
                   limit: Optional[int] = None, skip: Optional[int] = None,
                   sort: Optional[str] = None) -> PagedResponse[Conversation]:
        params: Dict[str, Any] = {}
        if type: params["type"] = type
        if q: params["q"] = q
        if limit: params["limit"] = limit
        if skip is not None: params["skip"] = skip
        if sort: params["sort"] = sort

        res = await self._http.request("GET", f"{self._v2}/team_conversations", params=params)
        return res.json()

    async def get(self, conv_id: str) -> Conversation:
        res = await self._http.request("GET", f"{self._v1}/team_conversations/{conv_id}")
        return res.json()

    async def get_by_conversation_id(self, conversation_id: str) -> Conversation:
        params = {"type": "conversation_id", "q": conversation_id}
        res = await self._http.request("GET", f"{self._v1}/team_conversations", params=params)
        return res.json()

    async def search(self, business_unit_id: str, q: str,
                     limit: Optional[int] = None, skip: Optional[int] = None) -> PagedResponse[Conversation]:
        params = {
            "business_unit_id": business_unit_id,
            "type": "text",
            "q": q,
        }
        if limit: params["limit"] = limit
        if skip is not None: params["skip"] = skip
        res = await self._http.request("GET", f"{self._v1}/team_conversations/_search", params=params)
        return res.json()

    async def get_views_count(self, type: Optional[str] = None, q: Optional[str] = None) -> Dict[str, int]:
        params = {}
        if type: params["type"] = type
        if q: params["q"] = q
        res = await self._http.request("GET", f"{self._v2}/team_conversations/_views_count", params=params)
        return res.json()

    async def get_outstanding(self, business_unit_id: str,
                              limit: Optional[int] = None, skip: Optional[int] = None) -> PagedResponse[Conversation]:
        params = {"type": "business_unit_id", "q": business_unit_id}
        if limit: params["limit"] = limit
        if skip is not None: params["skip"] = skip
        res = await self._http.request("GET", f"{self._v1}/team_conversations/_outstanding", params=params)
        return res.json()

    async def join(self, body: JoinConversationInput) -> ConversationActionResponse:
        res = await self._http.request("POST", f"{self._v1}/team_conversations/_join", json=body)
        return res.json()

    async def leave(self, body: LeaveConversationInput) -> ConversationActionResponse:
        res = await self._http.request("POST", f"{self._v1}/team_conversations/_leave", json=body)
        return res.json()

    async def update_status(self, body: UpdateStatusInput) -> ConversationActionResponse:
        res = await self._http.request("POST", f"{self._v1}/team_conversations/_update_status", json=body)
        return res.json()

    async def update_name(self, body: UpdateNameInput) -> ConversationActionResponse:
        res = await self._http.request("POST", f"{self._v1}/team_conversations/_update_name", json=body)
        return res.json()

    async def init_video_call(self, body: InitVideoCallInput) -> ConversationActionResponse:
        res = await self._http.request("POST", f"{self._v1}/team_conversations/_init_jaas_conference", json=body)
        return res.json()

    async def assign_team_member(self, body: AssignTeamMemberInput) -> ConversationActionResponse:
        res = await self._http.request("POST", f"{self._v1}/team_conversations/assign_team_member", json=body)
        return res.json()

    async def remove_team_member(self, body: RemoveTeamMemberInput) -> ConversationActionResponse:
        res = await self._http.request("POST", f"{self._v1}/team_conversations/remove_team_member", json=body)
        return res.json()

    async def get_invitable_users(self, team_conv_id: str) -> List[InvitableUser]:
        res = await self._http.request("GET", f"{self._v1}/team_conversations/{team_conv_id}/users")
        return res.json()

    async def get_conversation(self, conversation_id: str) -> Conversation:
        res = await self._http.request("GET", f"{self._v1}/conversations/{conversation_id}")
        return res.json()

    async def create(self, body: Optional[CreateConversationInput] = None) -> Conversation:
        res = await self._http.request("POST", f"{self._v1}/conversations", json=body or {})
        return res.json()

    async def join_request(self, body: JoinRequestInput) -> ConversationActionResponse:
        res = await self._http.request("POST", f"{self._v1}/team_conversations/_join_request", json=body)
        return res.json()

File: ts/src/resources/conversations.ts
import { HttpTransport } from "../http.js"
import type { Conversation, PagedResponse } from "../types/index.js"

// ─── Conversation action interfaces ──────────────────────────────────────────

export interface ConversationActionResponse {
  success: boolean
  conversation?: Conversation
  [key: string]: unknown
}

export interface JoinConversationInput {
  conversation_id: string
  [key: string]: unknown
}

export interface LeaveConversationInput {
  conversation_id: string
  [key: string]: unknown
}

export interface UpdateStatusInput {
  conversation_id: string
  status: string
  [key: string]: unknown
}

export interface UpdateNameInput {
  conversation_id: string
  name: string
  [key: string]: unknown
}

export interface InitVideoCallInput {
  conversation_id: string
  [key: string]: unknown
}

export interface AssignTeamMemberInput {
  conversation_id: string
  user_id: string
  [key: string]: unknown
}

export interface RemoveTeamMemberInput {
  conversation_id: string
  user_id: string
  [key: string]: unknown
}

export interface InvitableUser {
  _id: string
  name?: string
  email?: string
  [key: string]: unknown
}

export interface CreateConversationInput {
  channel_id?: string
  contact_id?: string
  [key: string]: unknown
}

export interface JoinRequestInput {
  conversation_id: string
  [key: string]: unknown
}

export class ConversationsResource {
  /**
   * @param base - channel-service base URL (gateway/channel-service)
   */
  constructor(private readonly http: HttpTransport, private readonly base: string) {}

  private get v1() { return `${this.base}/v1` }
  private get v2() { return `${this.base}/v2` }

  // ─── Team Conversations ──────────────────────────────────────────────────────

  async list(params?: {
    type?: string
    q?: string
    limit?: number
    skip?: number
    sort?: string
  }): Promise<PagedResponse<Conversation>> {
    const url = new URL(`${this.v2}/team_conversations`)
    if (params?.type)   url.searchParams.set("type",  params.type)
    if (params?.q)      url.searchParams.set("q",     params.q)
    if (params?.limit)  url.searchParams.set("limit", String(params.limit))
    if (params?.skip !== undefined) url.searchParams.set("skip", String(params.skip))
    if (params?.sort)   url.searchParams.set("sort",  params.sort)
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async get(convId: string): Promise<Conversation> {
    return this.http.getFetch()(`${this.v1}/team_conversations/${convId}`, { method: "GET" }).then(r => r.json())
  }

  async getByConversationId(conversationId: string): Promise<Conversation> {
    const url = new URL(`${this.v1}/team_conversations`)
    url.searchParams.set("type", "conversation_id")
    url.searchParams.set("q",    conversationId)
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async search(params: { businessUnitId: string; q: string; limit?: number; skip?: number }): Promise<PagedResponse<Conversation>> {
    const url = new URL(`${this.v1}/team_conversations/_search`)
    url.searchParams.set("business_unit_id", params.businessUnitId)
    url.searchParams.set("type",             "text")
    url.searchParams.set("q",               params.q)
    if (params.limit) url.searchParams.set("limit", String(params.limit))
    if (params.skip !== undefined) url.searchParams.set("skip", String(params.skip))
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async getViewsCount(params?: { type?: string; q?: string }): Promise<{ [view: string]: number }> {
    const url = new URL(`${this.v2}/team_conversations/_views_count`)
    if (params?.type) url.searchParams.set("type", params.type)
    if (params?.q)    url.searchParams.set("q",    params.q)
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async getOutstanding(params: { businessUnitId: string; limit?: number; skip?: number }): Promise<PagedResponse<Conversation>> {
    const url = new URL(`${this.v1}/team_conversations/_outstanding`)
    url.searchParams.set("type", "business_unit_id")
    url.searchParams.set("q",    params.businessUnitId)
    if (params.limit) url.searchParams.set("limit", String(params.limit))
    if (params.skip !== undefined) url.searchParams.set("skip", String(params.skip))
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async join(body: JoinConversationInput): Promise<ConversationActionResponse> {
    return this.http.getFetch()(`${this.v1}/team_conversations/_join`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async leave(body: LeaveConversationInput): Promise<ConversationActionResponse> {
    return this.http.getFetch()(`${this.v1}/team_conversations/_leave`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async updateStatus(body: UpdateStatusInput): Promise<ConversationActionResponse> {
    return this.http.getFetch()(`${this.v1}/team_conversations/_update_status`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async updateName(body: UpdateNameInput): Promise<ConversationActionResponse> {
    return this.http.getFetch()(`${this.v1}/team_conversations/_update_name`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async initVideoCall(body: InitVideoCallInput): Promise<ConversationActionResponse> {
    return this.http.getFetch()(`${this.v1}/team_conversations/_init_jaas_conference`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async assignTeamMember(body: AssignTeamMemberInput): Promise<ConversationActionResponse> {
    return this.http.getFetch()(`${this.v1}/team_conversations/assign_team_member`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async removeTeamMember(body: RemoveTeamMemberInput): Promise<ConversationActionResponse> {
    return this.http.getFetch()(`${this.v1}/team_conversations/remove_team_member`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async getInvitableUsers(teamConvId: string): Promise<InvitableUser[]> {
    return this.http.getFetch()(`${this.v1}/team_conversations/${teamConvId}/users`, { method: "GET" }).then(r => r.json())
  }

  // ─── Single Conversation ─────────────────────────────────────────────────────

  async getConversation(conversationId: string): Promise<Conversation> {
    return this.http.getFetch()(`${this.v1}/conversations/${conversationId}`, { method: "GET" }).then(r => r.json())
  }

  // ─── Create standalone conversation ─────────────────────────────────────────

  async create(body?: CreateConversationInput): Promise<Conversation> {
    return this.http.getFetch()(`${this.v1}/conversations`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body ?? {}),
    }).then(r => r.json())
  }

  async joinRequest(body: JoinRequestInput): Promise<ConversationActionResponse> {
    return this.http.getFetch()(`${this.v1}/team_conversations/_join_request`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }
}
